From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Tue, 20 May 2025 01:12:10 +0200
Subject: [PATCH] Create PreAPITeleportEvent

As spigot api teleportation is subject to stricter checks than the
default minecraft teleportation, some plugins may not be able to
teleport entities that would be perfectly fine teleporting under the
vanilla standart.

To not remove the created rules by spigot and potentially break plugin
compatibility the PreAPITeleportEvent is called prior to most checks
executed by spigot which allows plugins to modify potential target
entities that are teleported by other plugins prior to the checks.

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index b38073628d3b1381ccc4e5c6c44b2b7ec8ba1273..fdfef69ad3e14eeb1d717bac8231e406b0309c3a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -296,6 +296,9 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         // Paper end
         Preconditions.checkArgument(location != null, "location cannot be null");
         location.checkFinite();
+        // KTP start - Call pre api teleport / move dead check above event to respect definition of event
+        new dev.lynxplay.ktp.event.server.PreAPITeleportEvent(this, this.getLocation(), location.clone(), cause).callEvent();
+        // KTP end
         // Paper start - Teleport passenger API
         Set<io.papermc.paper.entity.TeleportFlag> flagSet = new HashSet<>(List.of(flags)); // Wrap into list while multiple old flags link to the same new one
         boolean dismount = !flagSet.contains(io.papermc.paper.entity.TeleportFlag.EntityState.RETAIN_VEHICLE);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 322a1e38799a7bf45b8d3ee151b0b62df45d55d7..da831b185204610810566d701e908c38d84d4322 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1398,6 +1398,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         if (this.getHealth() == 0 || entity.isRemoved()) {
             return false;
         }
+        // KTP start - Call pre api teleport
+        new dev.lynxplay.ktp.event.server.PreAPITeleportEvent(this, this.getLocation(), location.clone(), cause).callEvent();
+        // KTP end
 
         if (entity.connection == null) {
             return false;
